webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/metrics/Scheduler.scala

```diff
@@ -0,0 +1,130 @@
+* distributed under the license agreement of your choice.
+* A "Related Module" means a set of sources files including their
+* documentation that, without modification of the Source Code, enables
+* supplementary functions or services in addition to those offered by
+* the Software.
+*
+* Rudder is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
+
+*
+*************************************************************************************
+*/
+
+package com.normation.rudder.metrics
+
+import java.util.concurrent.TimeUnit
+
+import com.normation.errors._
+import zio._
+import zio.clock.Clock
+import zio.duration.Duration
+import zio.syntax._
+
+
+/*
+ * Scheduler to write logs in file. We want to log a new status line when interesting
+ * events happens, like a node is added or deleted, but not too often (once every
+ * 10min is well enought: you can't really use rudder below that timespan), and
+ * if nothing interesting happens, we still want to have a log line every once and
+ * then (1h).
+ * So we define a scheduler that calls a consumer (the actual logger) at most every
+ * `min` time and at least every `max` time, with:
+ * - min must be smaller than max strictly.
+ * Consumer must handle its errors.
+ * It uses a queue with one element only to notice not yet processed element.
+ *
+ * It should be created with `Scheduler.make` which check for parameters consistancy.
+ */
+object Scheduler {
+  /**
+   * Create a scheduler
+   * @param min: time to wait AT LEAST between action
+   * @param max: time to wait AT MOST between action
+   * @param action: action to execute periodically
+   * @param default: default value to use as parameter for `action` if it's not started by an event
+   * @param zclock: clock to use in that cron
+   */
+  def make[A](min: Duration, max: Duration, action: A => UIO[Unit], default: UIO[A], zclock: Clock): IOResult[Scheduler[A]] = {
+    if(min >= max) {
+      Inconsistency(s"Scheduler maximum period (${max.render}) cannot be shorter than its minimum period (${min.render})").fail
+    } else {
+      Queue.dropping[A](1).map(q => new Scheduler(min, max, q, action, default, zclock))
+    }
+  }
+}
+
+/*
+ * The actual class. Assume all parameters are OK. You should use `Scheduler.make` to create it.
+ */
+class Scheduler[A](
+    min    : Duration
+  , max    : Duration
+  , queue  : Queue[A]
+  , action : A => UIO[Unit]
+  , default: UIO[A] // the default value to used for consumer started by cron, not event
+  , zclock : Clock
+) {
+
+  /**
+   * External API to use when an event happens.
+   */
+  def triggerSchedule(a: A): UIO[Unit] = {
```

Comment:
There is nothing to trigger a Schedule ? is it expected ? 

Nonetheless, it's not needed for this feature and we can add it after.

Reply:
I wasn't sure about what exactly we wanted. Likely for added node, and deleted?
But I wanted to get the core feature in before validating what events we would like to add.